Add `addCacheCalls` in `BaseCodegenIntegrationSpec`

I'm hitting a bug here that needs to be resolved first. 

For a clean solution I have some ideas how to change the traversal logic that need to be discussed with @joroKr21.

```
assertion failed: BindingDef LH type `group.values.type` is not a supertype of RH type `org.emmalanguage.api.DataBag[org.emmalanguage.test.schema.Movies.ImdbMovie]`.
(lhs: `value values$r1`, rhs:
`LocalOps.cache[org.emmalanguage.test.schema.Movies.ImdbMovie](values$r1$r1)`
)
java.lang.AssertionError: assertion failed: BindingDef LH type `group.values.type` is not a supertype of RH type `org.emmalanguage.api.DataBag[org.emmalanguage.test.schema.Movies.ImdbMovie]`.
(lhs: `value values$r1`, rhs:
`LocalOps.cache[org.emmalanguage.test.schema.Movies.ImdbMovie](values$r1$r1)`
)
	at scala.Predef$.assert(Predef.scala:170)
	at org.emmalanguage.ast.Bindings$BindingAPI$BindingDef$.apply(Bindings.scala:107)
	at org.emmalanguage.ast.Values$ValueAPI$ValDef$.apply(Values.scala:101)
	at org.emmalanguage.compiler.lang.backend.Caching$Caching$.org$emmalanguage$compiler$lang$backend$Caching$Caching$$cacheAs(Caching.scala:54)
	at org.emmalanguage.compiler.lang.backend.Caching$Caching$$anonfun$addCacheCalls$1$$anonfun$apply$2$$anonfun$5.apply(Caching.scala:123)
	at org.emmalanguage.compiler.lang.backend.Caching$Caching$$anonfun$addCacheCalls$1$$anonfun$apply$2$$anonfun$5.apply(Caching.scala:117)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
...
```

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/Caching.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/Caching.scala
@@ -36,23 +36,23 @@ private[backend] trait Caching extends Common {
     import Core.{Lang => core}
     import UniverseImplicits._
 
+    private type CacheFlags = Map[u.Symbol, (Boolean, Int)]
+    private type CacheStore = Map[u.Symbol, u.ValDef]
+
     private val runtime = Some(Ops.ref)
     private val cache   = Ops.cache
-
-    /** Is `sym` a method defining a loop or the body of a loop? */
-    private def isLoop(sym: u.Symbol) = is.method(sym) &&
-      (api.Sym.findAnn[loop](sym).isDefined || api.Sym.findAnn[loopBody](sym).isDefined)
 
     /** Does `sym` have a type of `DataBag` or a subtype? */
     private def isDataBag(sym: u.Symbol) =
       api.Type.constructor(sym.info) =:= DataBag.tpe
 
-    /** Caches `x` as the new value `y`. */
-    private def cacheAs(x: u.TermSymbol, y: u.TermSymbol) = {
-      val targs = Seq(api.Type.arg(1, x.info))
-      val argss = Seq(Seq(core.Ref(x)))
-      core.ValDef(y, core.DefCall(runtime, cache, targs, argss))
-    }
+    /** Is `sym` a method defining a loop or the body of a loop? */
+    private def isLoop(sym: u.Symbol) = is.method(sym) &&
+      (api.Sym.findAnn[loop](sym).isDefined || api.Sym.findAnn[loopBody](sym).isDefined)
+
+    /** Is `sym` a method defining a suffix continuation? */
+    private def isSuffix(sym: u.Symbol) = is.method(sym) &&
+      api.Sym.findAnn[suffix](sym).isDefined
 
     /**
      * Wraps `DataBag` term in `cache(...)` calls if needed.
@@ -73,65 +73,66 @@ private[backend] trait Caching extends Common {
      */
     lazy val addCacheCalls: u.Tree => u.Tree = tree => {
       // Per DataBag reference, collect flag for access in a loop and ref count.
-      val refs = api.TopDown.withOwnerChain
-        .accumulateWith[Map[u.Symbol, (Boolean, Int)]] {
-          // Increment counter and set flag if referenced in a loop.
-          case Attr.inh(core.BindingRef(target), owners :: _) if isDataBag(target) =>
-            val inLoop = owners.reverseIterator.takeWhile(_ != target.owner).exists(isLoop)
-            Map(target -> (inLoop, 1))
-          // Arguments to loop methods are considered as referenced in the loop.
-          case Attr.none(core.DefCall(None, method, _, argss)) if isLoop(method) =>
-            argss.flatten.collect { case core.Ref(target) if isDataBag(target) =>
-              target -> (true, 1)
-            } (breakOut)
-        } (merge(catsKernelStdMonoidForTuple2(disj, implicitly))).traverseAny._acc(tree).head
+      val refs = api.TopDown.withOwnerChain.accumulateWith[CacheFlags] {
+        // Increment counter and set flag if referenced in a loop.
+        case Attr.inh(core.BindingRef(target), owners :: _) if isDataBag(target) =>
+          val inLoop = owners.reverseIterator.takeWhile {
+            owner => owner != target.owner && !isSuffix(owner)
+          }.exists(isLoop)
+          Map(target -> (inLoop, 1))
+        // Arguments to loop methods are considered as referenced in the loop.
+        case Attr.none(core.DefCall(None, method, _, argss)) if isLoop(method) =>
+          argss.flatten.collect { case core.Ref(target) if isDataBag(target) =>
+            target -> (true, 1)
+          } (breakOut)
+      } (merge(catsKernelStdMonoidForTuple2(disj, implicitly))).traverseAny._acc(tree).head
 
       /** Cache when referenced in a loop or more than once. */
-      def shouldCache(sym: u.Symbol) = {
-        val (inLoop, refCount) = refs(sym)
+      def shouldCache(x: u.Symbol) = {
+        val (inLoop, refCount) = refs(x)
         inLoop || refCount > 1
       }
 
       /** Creates cached values for parameters of methods and lambdas. */
-      def cacheParams(params: Seq[u.ValDef]): Map[u.Symbol, u.ValDef] =
-        params.collect { case core.ParDef(p, _) if shouldCache(p) =>
-          p -> cacheAs(p, api.ValSym(p.owner, api.TermName.fresh(p), p.info))
+      def cacheBindings(binds: Seq[u.ValDef]): CacheStore =
+        binds.collect { case core.BindingDef(x, _) if shouldCache(x) =>
+          val y = api.ValSym(x.owner, api.TermName.fresh(x), x.info.widen)
+          val targs = Seq(api.Type.arg(1, x.info))
+          val argss = Seq(Seq(core.Ref(x)))
+          x -> core.ValDef(y, core.DefCall(runtime, cache, targs, argss))
         } (breakOut)
 
-      api.TopDown.withParent.accumulate {
-        // Accumulate cached lambda and method parameters.
+      api.TopDown.withParent.accumulate[CacheStore] {
         case core.Lambda(_, params, _) =>
-          cacheParams(params)
+          cacheBindings(params)
         case core.DefDef(method, _, paramss, _) if !isLoop(method) =>
-          cacheParams(paramss.flatten)
+          cacheBindings(paramss.flatten)
+        case core.Let(vals, _, _) =>
+          cacheBindings(vals.filter {
+            case core.ValDef(_, core.DefCall(_, m, _, _)) => !API.DataBag$.ops(m)
+            case _ => true
+          })
       } (overwrite).transformWith {
+        // Prepend cached parameters, replace cached values.
         case Attr(let @ core.Let(vals, defs, expr), cached :: _, parent :: _, _) =>
-          val params = parent match {
-            case Some(core.Lambda(_, params, _)) => params
-            case Some(core.DefDef(_, _, paramss, _)) => paramss.flatten
+          val cachedParams = (parent match {
+            case Some(core.Lambda(_, ps, _)) => ps
+            case Some(core.DefDef(_, _, pss, _)) => pss.flatten
             case _ => Seq.empty
-          }
+          }).map(_.symbol).collect(cached)
 
-          // Prepend cached parameters, replace cached vals.
-          val cachedParams = params.map(_.symbol).collect(cached)
-          val cachedVals = vals.flatMap {
-            // Don't cache reads.
-            case value @ core.ValDef(_, core.DefCall(_, method, _, _))
-              if API.DataBag$.ops(method) => Seq(value)
-            case core.ValDef(x, rhs) if shouldCache(x) =>
-              val y = api.TermSym.fresh(x)
-              Seq(core.ValDef(y, rhs), cacheAs(y, x))
-            case value => Seq(value)
+          val cachedVals = vals.flatMap { v =>
+            v +: cached.get(v.symbol).toSeq
           }
 
           if (cachedParams.isEmpty && cachedVals.size == vals.size) let
           else core.Let(cachedParams ++ cachedVals, defs, expr)
 
         // Dont't replace references in cache(...).
-        case Attr.inh(tree, Some(call) :: _) if call.symbol == cache =>
-          tree
+        case Attr.inh(t, Some(call) :: _)
+          if call.symbol == cache => t
 
-        // Replace references to cached parameters.
+        // Replace references to cached parameters or values.
         case Attr.acc(core.Ref(x), cached :: _) if cached.contains(x) =>
           core.Ref(cached(x).symbol.asTerm)
       }._tree(tree)

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/Comprehension.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/Comprehension.scala
@@ -258,6 +258,14 @@ trait Comprehension extends Common
     def desugar(monad: u.Symbol): u.Tree => u.Tree =
       ReDeSugar.desugar(monad)
 
+    /** `resugar` the default [[DataBag]] monad. */
+    lazy val resugarDataBag: u.Tree => u.Tree =
+      ReDeSugar.resugar(API.DataBag.sym)
+
+    /** `desugar` the default [[DataBag]] monad. */
+    lazy val desugarDataBag: u.Tree => u.Tree =
+      ReDeSugar.desugar(API.DataBag.sym)
+
     // -------------------------------------------------------------------------
     // Normalize API
     // -------------------------------------------------------------------------
@@ -265,6 +273,10 @@ trait Comprehension extends Common
     /** Delegates to [[Normalize.normalize()]]. */
     def normalize(monad: u.Symbol)(tree: u.Tree): u.Tree =
       Normalize.normalize(monad)(tree)
+
+    /** `normalize` the default [[DataBag]] monad. */
+    lazy val normalizeDataBag: u.Tree => u.Tree =
+      Normalize.normalize(API.DataBag.sym)
 
     // -------------------------------------------------------------------------
     // Combine API

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/ReDeSugar.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/ReDeSugar.scala
@@ -174,7 +174,7 @@ private[comprehension] trait ReDeSugar extends Common {
           val gen = cs.Generator(genSym, hd)
           val hd1 = cs.Head(core.Let(expr = core.Ref(genSym)))
           core.Let(vals, defs, cs.Comprehension(qs :+ gen, hd1))
-      }._tree.andThen(Core.flatten)
+      }._tree.andThen(Core.unnest)
     }
   }
 }

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/ANF.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/ANF.scala
@@ -155,7 +155,7 @@ private[core] trait ANF extends Common {
           }
 
           src.Block(stats ++ inner, expr)
-      }._tree
+      }._tree.andThen(api.Owner.atEncl)
 
     /**
      * Converts a tree into administrative normal form (ANF).
@@ -180,7 +180,7 @@ private[core] trait ANF extends Common {
      * == Postconditions ==
      * - An ANF tree where all nested blocks have been flattened.
      */
-    lazy val flatten: u.Tree => u.Tree =
+    lazy val unnest: u.Tree => u.Tree =
       api.BottomUp.transform {
         case parent @ core.Let(vals, defs, expr) if hasNestedLets(parent) =>
           // Flatten nested let expressions in value position without control flow.
@@ -208,7 +208,7 @@ private[core] trait ANF extends Common {
 
           val (trimmedVals, trimmedExpr) = trimVals(flatVals ++ exprVals, flatExpr)
           core.Let(trimmedVals, flatDefs, trimmedExpr)
-      }.andThen(_.tree)
+      }._tree.andThen(api.Owner.atEncl)
 
     // ---------------
     // Helper methods

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/Core.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/Core.scala
@@ -316,8 +316,8 @@ trait Core extends Common
     /** Delegates to [[ANF.transform]]. */
     lazy val anf = ANF.transform
 
-    /** Delegates to [[ANF.flatten]]. */
-    lazy val flatten = ANF.flatten
+    /** Delegates to [[ANF.unnest]]. */
+    lazy val unnest = ANF.unnest
 
     /** Delegates to [[DSCF.stripAnnotations]]. */
     lazy val stripAnnotations = DSCF.stripAnnotations

--- a/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/Core.scala
+++ b/emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/Core.scala
@@ -303,8 +303,8 @@ trait Core extends Common
     /** Lifting. The canonical compiler frontend. */
     lazy val lift: u.Tree => u.Tree = Function.chain(Seq(
       lnf,
-      Comprehension.resugar(DataBag.sym),
-      Comprehension.normalize(DataBag.sym)
+      Comprehension.resugarDataBag,
+      Comprehension.normalizeDataBag
     ))
 
     /** Chains [[ANF.transform]], and [[DSCF.transform]]. */

--- a/emma-language/src/test/scala/org/emmalanguage/compiler/BaseCodegenIntegrationSpec.scala
+++ b/emma-language/src/test/scala/org/emmalanguage/compiler/BaseCodegenIntegrationSpec.scala
@@ -57,6 +57,7 @@ abstract class BaseCodegenIntegrationSpec extends BaseCompilerSpec with BeforeAn
       Core.lift,
       Optimizations.foldFusion,
       Comprehension.combine,
+      Backend.addCacheCalls,
       backendPipeline,
       addContext
     ).compose(_.tree)

--- a/emma-language/src/test/scala/org/emmalanguage/compiler/lang/backend/CachingSpec.scala
+++ b/emma-language/src/test/scala/org/emmalanguage/compiler/lang/backend/CachingSpec.scala
@@ -114,5 +114,21 @@ class CachingSpec extends BaseCompilerSpec {
 
       addCacheCalls(inp) shouldBe alphaEqTo(liftPipeline(exp))
     }
+
+    "not referenced once in the suffix of a loop method" in {
+      val inp = u.reify {
+        val xs = DataBag(1 to 100)
+        val ys = xs.map(x => x * x)
+        var i = 0
+        while (i < 100) {
+          println(i)
+          i *= i
+        }
+
+        ys
+      }
+
+      addCacheCalls(inp) shouldBe alphaEqTo(liftPipeline(inp))
+    }
   }
 }

--- a/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/CombinationSpec.scala
+++ b/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/CombinationSpec.scala
@@ -44,7 +44,7 @@ class CombinationSpec extends BaseCompilerSpec {
     pipeline(typeCheck = true)(
       Core.lnf,
       tree => time(Comprehension.combine(tree), "combine"),
-      Core.flatten
+      Core.unnest
     ).compose(_.tree)
 
   def applyOnce(rule: (u.Symbol, u.Tree) => Option[u.Tree]): u.Expr[Any] => u.Tree = {
@@ -55,7 +55,7 @@ class CombinationSpec extends BaseCompilerSpec {
     pipeline(typeCheck = true)(
       Core.lnf,
       tree => time(transform(tree), "match rule"),
-      Core.flatten
+      Core.unnest
     ).compose(_.tree)
   }
 

--- a/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/NormalizeSpec.scala
+++ b/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/NormalizeSpec.scala
@@ -37,14 +37,14 @@ class NormalizeSpec extends BaseCompilerSpec {
   val resugarPipeline: u.Expr[Any] => u.Tree =
     pipeline(typeCheck = true)(
       Core.lnf,
-      Comprehension.resugar(API.DataBag.sym)
+      Comprehension.resugarDataBag
     ).compose(_.tree)
 
   val normalizePipeline: u.Expr[Any] => u.Tree =
     pipeline(typeCheck = true)(
       Core.lnf,
-      Comprehension.resugar(API.DataBag.sym),
-      tree => time(Comprehension.normalize(API.DataBag.sym)(tree), "normalize")
+      Comprehension.resugarDataBag,
+      tree => time(Comprehension.normalizeDataBag(tree), "normalize")
     ).compose(_.tree)
 
   // ---------------------------------------------------------------------------

--- a/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/ReDeSugarSpec.scala
+++ b/emma-language/src/test/scala/org/emmalanguage/compiler/lang/comprehension/ReDeSugarSpec.scala
@@ -36,21 +36,17 @@ class ReDeSugarSpec extends BaseCompilerSpec {
       Core.lnf
     ).compose(_.tree)
 
-  val resugarPipeline: u.Expr[Any] => u.Tree = {
-    val resugar = Comprehension.resugar(API.DataBag.sym)
+  val resugarPipeline: u.Expr[Any] => u.Tree =
     pipeline(typeCheck = true)(
       Core.lnf,
-      tree => time(resugar(tree), "resugar")
+      tree => time(Comprehension.resugarDataBag(tree), "resugar")
     ).compose(_.tree)
-  }
 
-  val desugarPipeline: u.Expr[Any] => u.Tree = {
-    val desugar = Comprehension.desugar(API.DataBag.sym)
+  val desugarPipeline: u.Expr[Any] => u.Tree =
     pipeline(typeCheck = true)(
       Core.lnf,
-      tree => time(desugar(tree), "desugar")
+      tree => time(Comprehension.desugarDataBag(tree), "desugar")
     ).compose(_.tree)
-  }
 
   // ---------------------------------------------------------------------------
   // Spec data: a collection of (desugared, resugared) trees

